src/hotspot/share/runtime/handshake.cpp

```diff
@@ -327,37 +290,36 @@ void HandshakeOperation::do_handshake(JavaThread* thread) {
   // It is no longer safe to refer to 'this' as the VMThread/Handshaker may have destroyed this operation
 }
 
-void Handshake::execute(HandshakeClosure* thread_cl) {
-  HandshakeOperation cto(thread_cl);
+void Handshake::execute(HandshakeClosure* hs_cl) {
+  HandshakeOperation cto(hs_cl, NULL);
   VM_HandshakeAllThreads handshake(&cto);
   VMThread::execute(&handshake);
 }
 
-bool Handshake::execute(HandshakeClosure* thread_cl, JavaThread* target) {
-  HandshakeOperation cto(thread_cl);
-  VM_HandshakeOneThread handshake(&cto, target);
-  VMThread::execute(&handshake);
-  return handshake.executed();
-}
-
-bool Handshake::execute_direct(HandshakeClosure* thread_cl, JavaThread* target) {
+void Handshake::execute(HandshakeClosure* hs_cl, JavaThread* target) {
   JavaThread* self = JavaThread::current();
-  HandshakeOperation op(thread_cl, /*is_direct*/ true);
+  HandshakeOperation op(hs_cl, target);
 
   jlong start_time_ns = os::javaTimeNanos();
 
   ThreadsListHandle tlh;
   if (tlh.includes(target)) {
-    target->set_handshake_operation(&op);
+    target->handshake_state()->add_operation(&op);
   } else {
     log_handshake_info(start_time_ns, op.name(), 0, 0, "(thread dead)");
-    return false;
+    return;
   }
 
-  HandshakeState::ProcessResult pr =  HandshakeState::_no_operation;
+  int executed_by_driver = 0;
```

Comment:
The thread executing this handshake operation, what the current thread is doesn't matter.
You can't use current threads resources or be dependent otherwise on it.

Exception being locking issues in JVM TI, where we are dependent that requester have locked JVM TI state lock for us, but we are not dependent that the current thread is the owner. So checking that the lock is held by requester doesn't matter for how is the 'driver'.